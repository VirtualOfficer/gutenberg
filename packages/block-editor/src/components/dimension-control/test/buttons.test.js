/**
 * External dependencies
 */
import { shallow, mount } from 'enzyme';
import { uniqueId } from 'lodash';

/**
 * Internal dependencies
 */
import DimensionButtons from '../buttons';

describe( 'DimensionButtons', () => {
	const onChangeHandler = jest.fn();

	const sizesTable = [
		{
			name: 'None',
			abbr: 'No',
			size: 0,
			slug: 'none',
		},
		{
			name: 'Small',
			abbr: 'S',
			size: 14,
			slug: 'small',
		},
		{
			name: 'Medium',
			abbr: 'M',
			size: 24,
			slug: 'medium',
		},
		{
			name: 'Large',
			abbr: 'L',
			size: 34,
			slug: 'large',
		}, {
			name: 'Extra Large',
			abbr: 'XL',
			size: 60,
			slug: 'xlarge',
		},
	];

	afterEach( () => {
		onChangeHandler.mockClear();
	} );

	it( 'renders correctly with defaults', () => {
		const wrapper = shallow(
			<DimensionButtons
				id={ uniqueId() }
				sizes={ sizesTable }
			/>
		);
		expect( wrapper ).toMatchSnapshot();
	} );

	describe( 'required props', () => {
		it( 'does not render without a "id" prop', () => {
			const wrapper = shallow(
				<DimensionButtons
					sizes={ sizesTable } // providethe other required prop
				/>
			);

			expect( wrapper.isEmptyRender() ).toBe( true );
		} );

		it( 'does not render without a "sizes" prop', () => {
			const wrapper = shallow(
				<DimensionButtons
					id={ uniqueId() } // providethe other required prop
				/>
			);

			expect( wrapper.isEmptyRender() ).toBe( true );
		} );
	} );

	describe( 'selected states', () => {
		it( 'renders a selected button state for the matching "currentSize" prop', () => {
			const size = 'medium';

			const wrapper = mount(
				<DimensionButtons
					id={ uniqueId() }
					sizes={ sizesTable }
					currentSize={ size }
				/>
			);

			const selectedButton = wrapper.find( 'button' ).filter( '.is-primary' );

			expect( selectedButton ).toHaveLength( 1 );
			expect( selectedButton.prop( 'value' ) ).toBe( size );
			expect( selectedButton.prop( 'aria-pressed' ) ).toBe( true );
		} );

		it( 'does not render a selected button if no matching "currentSize" prop is provided', () => {
			const wrapper = mount(
				<DimensionButtons
					id={ uniqueId() }
					sizes={ sizesTable }
					currentSize="invalidsizehere"
				/>
			);

			const unpressedButtons = wrapper.find( 'button' ).filterWhere( ( button ) => {
				return false === button.prop( 'aria-pressed' );
			} );

			const pressedButtons = wrapper.find( 'button' ).filterWhere( ( button ) => {
				return true === button.prop( 'aria-pressed' );
			} );

			expect( wrapper.find( 'button' ).filter( '.is-primary' ) ).toHaveLength( 0 );
			expect( pressedButtons ).toHaveLength( 0 );
			expect( unpressedButtons ).toHaveLength( sizesTable.length );
		} );
	} );

	describe( 'callbacks', () => {
		it( 'triggers onChangeSpacingSize callback on button click', () => {
			const wrapper = mount(
				<DimensionButtons
					id={ uniqueId() }
					sizes={ sizesTable }
					currentSize="medium"
					onChangeSpacingSize={ onChangeHandler }
				/>
			);

			wrapper.find( 'button[value="small"]' ).simulate( 'click' );

			expect( onChangeHandler ).toHaveBeenCalledTimes( 1 );
		} );
	} );
} );
