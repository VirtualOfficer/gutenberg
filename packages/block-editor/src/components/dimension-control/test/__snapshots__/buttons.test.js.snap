// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DimensionButtons renders correctly with defaults 1`] = `
<Fragment>
  <ButtonGroup
    className="block-editor-dimension-control__buttons"
    id="block-editor-dimension-control-buttons-1"
  >
    <Tooltip
      key="none"
      text="None"
    >
      <ForwardRef(Button)
        aria-pressed={false}
        className="block-editor-dimension-control__button"
        isDefault={true}
        isPrimary={false}
        value="none"
      >
        <abbr
          title="None"
        >
          No
        </abbr>
      </ForwardRef(Button)>
    </Tooltip>
    <Tooltip
      key="small"
      text="Small"
    >
      <ForwardRef(Button)
        aria-pressed={false}
        className="block-editor-dimension-control__button"
        isDefault={true}
        isPrimary={false}
        value="small"
      >
        <abbr
          title="Small"
        >
          S
        </abbr>
      </ForwardRef(Button)>
    </Tooltip>
    <Tooltip
      key="medium"
      text="Medium"
    >
      <ForwardRef(Button)
        aria-pressed={false}
        className="block-editor-dimension-control__button"
        isDefault={true}
        isPrimary={false}
        value="medium"
      >
        <abbr
          title="Medium"
        >
          M
        </abbr>
      </ForwardRef(Button)>
    </Tooltip>
    <Tooltip
      key="large"
      text="Large"
    >
      <ForwardRef(Button)
        aria-pressed={false}
        className="block-editor-dimension-control__button"
        isDefault={true}
        isPrimary={false}
        value="large"
      >
        <abbr
          title="Large"
        >
          L
        </abbr>
      </ForwardRef(Button)>
    </Tooltip>
    <Tooltip
      key="xlarge"
      text="Extra Large"
    >
      <ForwardRef(Button)
        aria-pressed={false}
        className="block-editor-dimension-control__button"
        isDefault={true}
        isPrimary={false}
        value="xlarge"
      >
        <abbr
          title="Extra Large"
        >
          XL
        </abbr>
      </ForwardRef(Button)>
    </Tooltip>
  </ButtonGroup>
</Fragment>
`;
